### PR TITLE
EA: perform analysis once for post-optimization IR, and remove `IPO EA`

### PR DESCRIPTION
**Pull request by aviatesk**
Following the discussions and changes in #50805, we now consider post-inlining IR as IPO-valid. Revisiting EA, I've realized that running EA twice—once for computing IPO-valid escape cache and once for local optimization analysis—is redundant. This commit streamlines the EA process to perform the analysis just once on post-optimization IR, and caches that result. This change also removes all interprocedural EA code, which had significant overlap with inlining code.

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
SGTM. But will this introduce any issues with convergence, or is that already addressed by #51092?

Also, do we need to be concerned with this that `return_type_tfunc` is marked `EFFECTS_TOTAL` and `MethodResultPure`, even though it would be a significant soundness issue to run code that contains that call via const-eval?

**Comment by aviatesk** [comment]
> But will this introduce any issues with convergence, or is that already addressed by #51092?

I don't foresee any convergence issues, as EA will be executed on each `optimize`-d frame and it won't be recursive. Although EA will cache and reuse results, it won't recurse down into a callee that hasn't been analyzed yet.

> Also, do we need to be concerned with this that `return_type_tfunc` is marked `EFFECTS_TOTAL` and `MethodResultPure`, even though it would be a significant soundness issue to run code that contains that call via const-eval?

This seems like a separate concern? Considering the effects modeling of that function, I'm actually uncertain why it's marked `EFFECTS_TOTAL`. The `return_type` call does have an externally observable effect, because it creates an inference cache.

**Review by vtjnash — APPROVED** [review]

**Comment by vtjnash** [comment]
> EA will be executed on each optimize-d frame and it won't be recursive. Although EA will cache and reuse results, it won't recurse down into a callee that hasn't been analyzed yet.

I guess that is fine. It adds another reason that the computed results from inference are not consistent, but we already have fundamental reasons of computability that the computed effects cannot be accessed in a consistent-preserving way. (And I maintain my hope that someday we can expose Core.Compiler.return_effect optimization in the same way as Core.Compiler.return_type :grin:)

File: test/compiler/EscapeAnalysis/EAUtils.jl [hunk 1]
```diff
@@ -192,51 +188,43 @@ function run_passes_ipo_safe_with_ea(interp::EscapeAnalyzer,
...
         @timeit "verify 3" (verify_ir(ir); verify_linetable(ir.linetable))
     end
+    nargs = let def = sv.linfo.def; isa(def, Method) ? Int(def.nargs) : 0; end
+    get_escape_cache = GetEscapeCache(interp)
+    local estate::EscapeState
+    try
+        @timeit "EA" estate = analyze_escapes(ir, nargs, get_escape_cache)
+    catch err
+        @error "error happened within EA, inspect `Main.ir` and `Main.nargs`"
+        @eval Main (ir = $ir; nargs = $nargs)
```
Comment (jpsamaroo) [inline test/compiler/EscapeAnalysis/EAUtils.jl:207]: Maybe while we keep this debugging logic around, we should give these more explanatory names?

```suggestion
        @error "error happened within EA, inspect `Main.failed_escapeanalysis_ir` and `Main.failed_escapeanalysis_nargs`"
        @eval Main (failed_escapeanalysis_ir = $ir; failed_escapeanalysis_nargs = $nargs)
```

**Pull request merged**